http-server/src/main/java/org/triplea/server/reporting/error/ErrorReportGateKeeper.java

```diff
@@ -0,0 +1,18 @@
+package org.triplea.server.reporting.error;
+
+import java.util.function.Predicate;
+
+/**
+ * An object to check if we should accept an error report request from our users and use that
+ * request to create a bug tracking ticket. A reason to deny is if rate limit has been exceeded.
+ */
+class ErrorReportGateKeeper implements Predicate<ErrorReportRequest> {
+
+  /** True if we should allow the error request. */
+  @Override
+  public boolean test(final ErrorReportRequest errorReportRequest) {
+    // TODO: implement
+    // TODO: log a warning when we do a request reject and return false
+    return true;
```

Comment:
A thing I haven't considered yet: If we're going full-HTTP/websocket, we might just be able to "hide" our server behind nginx (reverse proxy, like we do it on the forum), and let nginx handle all the rate limiting + HTTPS for us.
This way less code for us to write, just some entries in the config file.
https://www.nginx.com/blog/rate-limiting-nginx/

Reply:
I agree nginx would be a good idea and should be used to throttle high rates of request.

We still may want this application level check for application level logic. It is a decision if we want nginx to handle *all* throttling. An error message of "You've submitted too many error reports, please wait" would perhaps be nicer than an outright reject request.  If not for the UX, at a maintenance and architcture level we  need to consider if we would want nginx to have application level knowledge at all and know which URLs should have which throttle rates.

Beyond that, we likely want an application level throttle at some point to prevent spam messages from being entered (ie: bad word check).

I can be persuaded otherwise, my current opinion is that nginx is probably better for global rules. The code to query DB is not necessarily complex, gives a better UX, and we likely need a level of application knowledge as well to check for additional reasons to discard requests. I think putting application level knowledge in nginx could become a bit awkward to manage.

Reply:
Good point.
No, having application logic inside nginx config is not a good idea.
I don't think nginx will simply close the connection on too many requests, but instead return 503 status or something, which we could handle in the client which is not the problem.
I do think route specific throttling should not be based of IPs or any other connection information though. Instead for chat messages for example, we should check the current users message limit and decrease it for every message this user sends.
This keeps our code rather simple while still getting benefits out of both worlds.